File: notale-v2/pptx2html-harness/agent.py
"""One Chat history and a direct tool loop. Domain delivery policy is a callback."""
import json
from pathlib import Path
import time

from model import MODEL
from tools import TOOLS


class Record:
    def __init__(self, directory):
        self.root = Path(directory)
        self.root.mkdir(parents=True, exist_ok=True)

    def save(self, name, value):
        (self.root/name).write_text(json.dumps(value, ensure_ascii=False, indent=2)+'\n')

    def log(self, kind, data):
        with (self.root/'events.jsonl').open('a') as f:
            f.write(json.dumps({'time': time.time(), 'type': kind, 'data': data}, ensure_ascii=False)+'\n')


class Agent:
    def __init__(self, model, tools, record, max_turns=None):
        self.model, self.tools, self.record, self.max_turns = model, tools, record, max_turns

    async def run(self, system, task, accept_final):
        history = [{'role': 'system', 'content': system}, {'role': 'user', 'content': task}]
        result = {'status': 'running', 'model': MODEL, 'turns': 0, 'usage': [], 'quality_evaluated': False}
        try:
            while self.max_turns is None or result['turns'] < self.max_turns:
                turn = result['turns']+1
                request = {'model': MODEL, 'messages': history, 'tools': TOOLS,
                           'tool_choice': 'auto', 'reasoning_effort': 'low', 'stream': False}
                self.record.save(f'request-{turn:03}.json', request)
                response = await self.model.complete(request)
                self.record.save(f'response-{turn:03}.json', response)
                result['turns'] = turn
                result['usage'].append({'response_id': response.get('id'), 'usage': response.get('usage')})
                choice = response['choices'][0]
                if choice.get('finish_reason') not in ('stop', 'tool_calls'):
                    raise RuntimeError(f'Model did not complete: {choice.get("finish_reason")}')
                message = choice['message']
                calls = message.get('tool_calls') or []
                if not calls and not message.get('content'):
                    raise RuntimeError('Empty model response; no synthetic completion')
                # Keep each full tool_call, including extra_content.google.thought_signature.
                # Only output-only assistant fields are omitted from the next request.
                history.append({k: message[k] for k in (
                    'role', 'content', 'tool_calls', 'reasoning', 'reasoning_content', 'thinking') if k in message})
                images = []
                for call in calls:
                    try:
                        fn = call['function']
                        value = await self.tools.call(fn['name'], json.loads(fn['arguments']), call['id'])
                        if fn['name'] == 'view_image':
                            images.append({'type': 'image_url', 'image_url': {'url': value['image_url']}})
                            value = {'viewed': json.loads(fn['arguments'])['path'], 'pixels_in_next_user_message': True}
                    except Exception as e:
                        value = {'error': str(e)}
                    history.append({'role': 'tool', 'tool_call_id': call['id'],
                                    'content': json.dumps(value, ensure_ascii=False)})
                if images:
                    history.append({'role': 'user', 'content': [
                        {'type': 'text', 'text': 'Images in this turn’s view_image call order:'}, *images]})
                if calls:
                    continue
                feedback = accept_final()
                if feedback:
                    history.append({'role': 'user', 'content': feedback})
                    continue
                result.update(status='delivered_unreviewed', final=message['content'])
                (self.record.root.parent/'final.md').write_text(message['content'])
                return result
            result['status'] = 'budget_exhausted'
            return result
        except BaseException as e:
            result.update(status='error', error=str(e))
            raise
        finally:
            try:
                self.record.save('result.json', result)
            finally:
                await self.tools.close()

File: notale-ts/src/core/agent.ts
import { appendFile } from "node:fs/promises";
import type { ChatMessage, ChatModel, ToolDefinition } from "../adapters/models/chat-model.js";
import { redact } from "./redact.js";

export interface AgentOptions {
  model: ChatModel;
  system: string;
  prompt: string;
  tools: ToolDefinition[];
  execute(name: string, args: Record<string, unknown>): Promise<string | AgentToolOutput>;
  traceFile: string;
  signal?: AbortSignal;
  maxTurns?: number;
}

export interface AgentToolOutput { text: string; images?: string[] }

export async function runAgent(options: AgentOptions): Promise<{ text: string; turns: number }> {
  const history: ChatMessage[] = [{ role: "system", content: options.system }, { role: "user", content: options.prompt }];
  const maximum = options.maxTurns ?? 14;
  for (let turn = 1; turn <= maximum; turn += 1) {
    if (options.signal?.aborted) throw new Error("Generation cancelled");
    const response = await options.model.respond(history, options.tools, options.signal);
    await appendFile(options.traceFile, `${redact(JSON.stringify({ timestamp: new Date().toISOString(), type: "assistant", requestId: response.id, message: response.message, usage: { inputTokens: response.inputTokens, outputTokens: response.outputTokens, cachedTokens: response.cachedTokens } }))}\n`, "utf8");
    history.push(response.message);
    const calls = response.message.tool_calls ?? [];
    if (!calls.length) return { text: typeof response.message.content === "string" ? response.message.content.trim() : "", turns: turn };
    for (const call of calls) {
      let output: string;
      let images: string[] = [];
      try {
        const parsed: unknown = JSON.parse(call.function.arguments || "{}");
        if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("arguments must be a JSON object");
        const result = await options.execute(call.function.name, parsed as Record<string, unknown>);
        output = typeof result === "string" ? result : result.text;
        images = typeof result === "string" ? [] : result.images ?? [];
      } catch (error) {
        output = `${call.function.name} 失败：${error instanceof Error ? error.message : String(error)}`;
      }
      output = redact(output);
      await appendFile(options.traceFile, `${JSON.stringify({ timestamp: new Date().toISOString(), type: "tool", requestId: response.id, callId: call.id, name: call.function.name, output })}\n`, "utf8");
      history.push({ role: "tool", tool_call_id: call.id, content: output });
      if (images.length) history.push({ role: "user", content: [
        { type: "text", text: `下面是 ${call.function.name} 返回的图片，请检查真实画面后再决定如何使用或修正。` },
        ...images.map((url) => ({ type: "image_url" as const, image_url: { url } })),
      ] });
    }
  }
  throw new Error(`Agent exceeded ${maximum} turns`);
}
